### PR TITLE
Elaborate a bit on the meaning of idempotency

### PR DESCRIPTION
**Pull request by smklein**
Following-up from our chat-based conversation - this is where I stumbled with the term "idempotency" originally, so I figured I'd update those corresponding docs.

### CODE REVIEW

**Review by andrewjstone — COMMENTED** [review]
Despite our discussion, this definition does not make sense to me. I fail to see how an idempotent operation cannot return the same result. I believe the issue is simply that creating a UUID is *not* an idempotent operation.

For steno to work, we must separate any non-idempotent operations, such as generating a UUID into the 'start' state. After that, none are allowed.

**Comment by smklein** [comment]
> Despite our discussion, this definition does not make sense to me. I fail to see how an idempotent operation cannot return the same result. I believe the issue is simply that creating a UUID is _not_ an idempotent operation.
> 
> For steno to work, we must separate any non-idempotent operations, such as generating a UUID into the 'start' state. After that, none are allowed.

What do you mean by "issue"? What do you mean by "for steno to work"?

Steno currently is used this way, and from what I can tell, this is the intended behavior. What failure mode do you think exists with the implementation today?

**Comment by smklein** [comment]
> Despite our discussion, this definition does not make sense to me. I fail to see how an idempotent operation cannot return the same result. I believe the issue is simply that creating a UUID is _not_ an idempotent operation.
> 
> For steno to work, we must separate any non-idempotent operations, such as generating a UUID into the 'start' state. After that, none are allowed.

Here's a non-steno related example of an "idempotent" method returning different results: the HTTP DELETE method.

Calling "DELETE" with a particular UUID will result in the same side-effect to the system (only that one object will be deleted, even if called repeatedly), bu the result of the request may change, depending on the state of the system. The first response may be "OK, object deleted", and subsequent responses my be "NOT FOUND", but the operation is considered "idempotent", within a certain context.

That's part of why I wanted to clarify this meaning - "idempotent" in a mathematical sense requires an identical return value, but in a programming sense, is often used differently.

**Comment by andrewjstone** [comment]
> Here's a non-steno related example of an "idempotent" method returning different results: the HTTP DELETE method.
> 
> Calling "DELETE" with a particular UUID will result in the same side-effect to the system (only that one object will be deleted, even if called repeatedly), bu the result of the request may change, depending on the state of the system. The first response may be "OK, object deleted", and subsequent responses my be "NOT FOUND", but the operation is considered "idempotent", within a certain context.
> 
> That's part of why I wanted to clarify this meaning - "idempotent" in a mathematical sense requires an identical return value, but in a programming sense, is often used differently.

That's a great example with regards to a return value being different with an idempotent operation. Thank you for the clarity here. My main objection is to saying that "creating a uuid" is idempotent. After this example, I see where you are coming from and don't actually think we are disagreeing philosophically. I'm not even sure changing your wording is needed. I just had a knee jerk reaction, because creation of a UUID is definitely not idempotent. It's the equivalent of a POST operation. Let me see if I can convey this better...
 
I think the disconnect is that creation of the UUID is only run once for any given action, which is why the action as a whole is logically idempotent. The UUID in this case is the resource, not the URI/key. The action is equivalent to a user creating a UUID then calling put for a different key (the action key) with the UUID as the value. So while an action creates a UUID, it can only be run once, and the recording (equivalent to PUT) is idempotent. The compare-and-swap nature of the action based on action id (I'm making this up will dig more into steno later today) is equivalent to an HTTP PUT/DELETE at a known resource.

That whole explanation though doesn't fly for saga creation though, because that is actually similar to a POST where the URI fragment is something like  `/saga/create` and a new resource is created each time. But using that resource, one can manipulate it with idempotent saga actions that are idempotent by nature of the fact that they only output resource information once.

Interestingly, I'm not actually sure if the whole saga is idempotent because accessing it each time generates new information in the form of UUIDS and side effects, and it may revert due to failure, but each action is idempotent. In short, does idempotence actually compose?

**Comment by andrewjstone** [comment]
I talked with @smklein offline about this. The wording is clear in what it's trying to get across and I'm fine leaving it as is. I find thinking about idempotence in terms of sagas to be "hard", because the actions a saga performs seem to be things you want to do once. Not sure if there is a better word, but no need to quibble.

**Review by andrewjstone — APPROVED** [review]

File: src/saga_action_generic.rs [hunk 1]
```diff
@@ -105,7 +105,18 @@ pub trait Action<UserType: SagaType>: Debug + Send + Sync {
...
      * Executes the action for this saga node, whatever that is.  Actions
      * function like requests in distributed sagas: critically, they must be
-     * idempotent.  They should be very careful in using interfaces outside of
+     * idempotent. This means that multiple calls to the action have the
+     * same result on the system as a single call, although the action is not
+     * necessarily required to return the same result.
+     *
+     * As an example, generating a UUID to represent an object is a common saga
+     * action: if called repeatedly, it may generate different results, but it
+     * has the no side effects on the rest of the system. Similarly, using a
```
Comment (davepacheco) [inline src/saga_action_generic.rs:114]:
```suggestion
     * has no side effects on the rest of the system. Similarly, using a
```

(may need to rewrap after that)

**Review by davepacheco — APPROVED** [review]

File: src/saga_action_generic.rs (hunk 1 above)
Comment (smklein) [reply src/saga_action_generic.rs]: Done

**Pull request merged**